packages/connection/src/Manager.php

```diff
@@ -278,4 +288,195 @@ public function build_connect_url( $raw, $redirect, $from, $register ) {
+	 *
+	 * Blog Tokens can be "Normal" or "Special".
+	 * * Normal: The result of a normal connection flow. They look like
+	 *   "{$random_string_1}.{$random_string_2}"
+	 *   That is, $token_key and $private are both random strings.
+	 *   Sites only have one Normal Blog Token. Normal Tokens are found in either
+	 *   Jetpack_Options::get_option( 'blog_token' ) (usual) or the JETPACK_BLOG_TOKEN
+	 *   constant (rare).
+	 * * Special: A connection token for sites that have gone through an alternative
+	 *   connection flow. They look like:
+	 *   ";{$special_id}{$special_version};{$wpcom_blog_id};.{$random_string}"
+	 *   That is, $private is a random string and $token_key has a special structure with
+	 *   lots of semicolons.
+	 *   Most sites have zero Special Blog Tokens. Special tokens are only found in the
+	 *   JETPACK_BLOG_TOKEN constant.
+	 *
+	 * In particular, note that Normal Blog Tokens never start with ";" and that
+	 * Special Blog Tokens always do.
+	 *
+	 * When searching for a matching Blog Tokens, Blog Tokens are examined in the following
+	 * order:
+	 * 1. Defined Special Blog Tokens (via the JETPACK_BLOG_TOKEN constant)
+	 * 2. Stored Normal Tokens (via Jetpack_Options::get_option( 'blog_token' ))
+	 * 3. Defined Normal Tokens (via the JETPACK_BLOG_TOKEN constant)
+	 *
+	 * @param int|false    $user_id   false: Return the Blog Token. int: Return that user's User Token.
+	 * @param string|false $token_key If provided, check that the token matches the provided input.
+	 *                                false                                : Use first token. Default.
+	 *                                Jetpack_Data::MAGIC_NORMAL_TOKEN_KEY : Use first Normal Token.
+	 *                                non-empty string                     : Use matching token
+	 * @return object|false
+	 */
+	public function get_access_token( $user_id = false, $token_key = false ) {
+		$possible_special_tokens = array();
+		$possible_normal_tokens  = array();
+
+		if ( $user_id ) {
+			if ( ! $user_tokens = $this->option_manager->get_option( 'user_tokens' ) ) {
+				return false;
+			}
+			if ( self::MASTER_USER === $user_id ) {
+				if ( ! $user_id = $this->option_manager->get_option( 'master_user' ) ) {
+					return false;
+				}
+			}
+			if ( ! isset( $user_tokens[ $user_id ] ) || ! $user_tokens[ $user_id ] ) {
+				return false;
+			}
+			$user_token_chunks = explode( '.', $user_tokens[ $user_id ] );
+			if ( empty( $user_token_chunks[1] ) || empty( $user_token_chunks[2] ) ) {
+				return false;
+			}
+			if ( $user_id != $user_token_chunks[2] ) {
+				return false;
+			}
+			$possible_normal_tokens[] = "{$user_token_chunks[0]}.{$user_token_chunks[1]}";
+		} else {
+			$stored_blog_token = $this->option_manager->get_option( 'blog_token' );
+			if ( $stored_blog_token ) {
+				$possible_normal_tokens[] = $stored_blog_token;
+			}
+
+			$defined_tokens = \Jetpack_Constants::is_defined( 'JETPACK_BLOG_TOKEN' )
```

Comment:
Should we be consuming `Jetpack_Constants` from the constants package instead?

Reply:
We are! It's class mapped!

Reply:
Nice, yeah, you're right! Yeah, seems like we need to land #12558 to make this more obvious 😉